File: android/src/main/java/com/t6x/plugins/lancedb/LanceDBPlugin.kt
package com.t6x.plugins.lancedb

import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import uniffi.lancedb_ffi.LanceDbHandle

@CapacitorPlugin(name = "LanceDB")
class LanceDBPlugin : Plugin() {

    private var handle: LanceDbHandle? = null
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    @PluginMethod
    fun open(call: PluginCall) {
        val rawPath = call.getString("dbPath")
            ?: return call.reject("dbPath is required")
        val embeddingDim = call.getInt("embeddingDim")
            ?: return call.reject("embeddingDim is required")

        // Resolve "files://<rel>" to the app's private files directory
        val dbPath = if (rawPath.startsWith("files://")) {
            val rel = rawPath.removePrefix("files://")
            "${context.filesDir.absolutePath}/$rel"
        } else {
            rawPath
        }

        scope.launch {
            try {
                handle = LanceDbHandle.open(dbPath, embeddingDim)
                call.resolve()
            } catch (e: Exception) {
                call.reject("Failed to open LanceDB: ${e.message}", e)
            }
        }
    }

    // ── Generic Vector DB API ──────────────────────────────────

    @PluginMethod
    fun store(call: PluginCall) { doStore(call) }

    @PluginMethod
    fun search(call: PluginCall) { doSearch(call) }

    @PluginMethod
    fun delete(call: PluginCall) { doDelete(call) }

    @PluginMethod
    fun list(call: PluginCall) { doList(call) }

    @PluginMethod
    fun clear(call: PluginCall) { doClear(call) }

    // ── Deprecated memory-prefixed aliases ─────────────────────

    @Deprecated("Use store() instead")
    @PluginMethod
    fun memoryStore(call: PluginCall) { doStore(call) }

    @Deprecated("Use search() instead")
    @PluginMethod
    fun memorySearch(call: PluginCall) { doSearch(call) }

    @Deprecated("Use delete() instead")
    @PluginMethod
    fun memoryDelete(call: PluginCall) { doDelete(call) }

    @Deprecated("Use list() instead")
    @PluginMethod
    fun memoryList(call: PluginCall) { doList(call) }

    @Deprecated("Use clear() instead")
    @PluginMethod
    fun memoryClear(call: PluginCall) { doClear(call) }

    // ── Private helpers ────────────────────────────────────────

    private fun doStore(call: PluginCall) {
        val h = handle ?: return call.reject("LanceDB not initialized — call open() first")
        val key = call.getString("key") ?: return call.reject("key is required")
        val agentId = call.getString("agentId") ?: return call.reject("agentId is required")
        val text = call.getString("text") ?: return call.reject("text is required")
        val embeddingArr = call.getArray("embedding")
            ?: return call.reject("embedding is required")
        val metadata = call.getString("metadata")

        val embedding = jsonArrayToFloatList(embeddingArr)

        scope.launch {
            try {
                h.store(key, agentId, text, embedding, metadata)
                call.resolve()
            } catch (e: Exception) {
                call.reject("store failed: ${e.message}", e)
            }
        }
    }

    private fun doSearch(call: PluginCall) {
        val h = handle ?: return call.reject("LanceDB not initialized — call open() first")
        val queryVectorArr = call.getArray("queryVector")
            ?: return call.reject("queryVector is required")
        val limit = call.getInt("limit", 5) ?: 5
        val filter = call.getString("filter")

        val queryVector = jsonArrayToFloatList(queryVectorArr)

        scope.launch {
            try {
                val results = h.search(queryVector, limit.toUInt(), filter)
                val arr = JSArray()
                for (r in results) {
                    val obj = JSObject()
                    obj.put("key", r.key)
                    obj.put("text", r.text)
                    obj.put("score", r.score)
                    if (r.metadata != null) {
                        obj.put("metadata", r.metadata)
                    }
                    arr.put(obj)
                }
                val ret = JSObject()
                ret.put("results", arr)
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("search failed: ${e.message}", e)
            }
        }
    }

    private fun doDelete(call: PluginCall) {
        val h = handle ?: return call.reject("LanceDB not initialized — call open() first")
        val key = call.getString("key") ?: return call.reject("key is required")

        scope.launch {
            try {
                h.delete(key)
                call.resolve()
            } catch (e: Exception) {
                call.reject("delete failed: ${e.message}", e)
            }
        }
    }

    private fun doList(call: PluginCall) {
        val h = handle ?: return call.reject("LanceDB not initialized — call open() first")
        val prefix = call.getString("prefix")
        val limit = call.getInt("limit")

        scope.launch {
            try {
                val keys = h.list(prefix, limit?.toUInt())
                val arr = JSArray()
                for (k in keys) {
                    arr.put(k)
                }
                val ret = JSObject()
                ret.put("keys", arr)
                call.resolve(ret)
            } catch (e: Exception) {
                call.reject("list failed: ${e.message}", e)
            }
        }
    }

    private fun doClear(call: PluginCall) {
        val h = handle ?: return call.reject("LanceDB not initialized — call open() first")
        val collection = call.getString("collection")

        scope.launch {
            try {
                h.clear(collection)
                call.resolve()
            } catch (e: Exception) {
                call.reject("clear failed: ${e.message}", e)
            }
        }
    }

    override fun handleOnDestroy() {
        scope.cancel()
        handle?.close()
        handle = null
    }

    private fun jsonArrayToFloatList(arr: JSArray): List<Float> {
        val list = mutableListOf<Float>()
        for (i in 0 until arr.length()) {
            list.add(arr.getDouble(i).toFloat())
        }
        return list
    }
}

File: ios/Sources/LanceDBPlugin/LanceDBPlugin.swift
import Foundation
import Capacitor

@objc(LanceDBPlugin)
public class LanceDBPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "LanceDBPlugin"
    public let jsName = "LanceDB"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "open", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "memoryStore", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "memorySearch", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "memoryDelete", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "memoryList", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "memoryClear", returnType: CAPPluginReturnPromise),
    ]

    private var handle: LanceDBHandle?
    private let queue = DispatchQueue(label: "com.t6x.lancedb", qos: .userInitiated, attributes: .concurrent)

    @objc func open(_ call: CAPPluginCall) {
        guard let dbPath = call.getString("dbPath") else {
            return call.reject("dbPath is required")
        }
        guard let embeddingDim = call.getInt("embeddingDim") else {
            return call.reject("embeddingDim is required")
        }

        Task {
            do {
                self.handle = try await LanceDBHandle.open(dbPath: dbPath, embeddingDim: Int32(embeddingDim))
                call.resolve()
            } catch {
                call.reject("Failed to open LanceDB: \(error.localizedDescription)")
            }
        }
    }

    @objc func memoryStore(_ call: CAPPluginCall) {
        guard let handle else { return call.reject("LanceDB not initialized — call open() first") }
        guard let key = call.getString("key") else { return call.reject("key is required") }
        guard let agentId = call.getString("agentId") else { return call.reject("agentId is required") }
        guard let text = call.getString("text") else { return call.reject("text is required") }
        guard let embeddingArr = call.getArray("embedding") as? [NSNumber] else {
            return call.reject("embedding is required")
        }

        let embedding = embeddingArr.map { $0.floatValue }
        let metadata = call.getString("metadata")

        Task {
            do {
                try await handle.store(key: key, agentId: agentId, text: text, embedding: embedding, metadata: metadata)
                call.resolve()
            } catch {
                call.reject("memoryStore failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func memorySearch(_ call: CAPPluginCall) {
        guard let handle else { return call.reject("LanceDB not initialized — call open() first") }
        guard let queryArr = call.getArray("queryVector") as? [NSNumber] else {
            return call.reject("queryVector is required")
        }

        let queryVector = queryArr.map { $0.floatValue }
        let limit = UInt32(call.getInt("limit") ?? 5)
        let filter = call.getString("filter")

        Task {
            do {
                let results = try await handle.search(queryVector: queryVector, limit: limit, filter: filter)
                var arr: [[String: Any]] = []
                for r in results {
                    var obj: [String: Any] = [
                        "key": r.key,
                        "text": r.text,
                        "score": r.score
                    ]
                    if let meta = r.metadata {
                        obj["metadata"] = meta
                    }
                    arr.append(obj)
                }
                call.resolve(["results": arr])
            } catch {
                call.reject("memorySearch failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func memoryDelete(_ call: CAPPluginCall) {
        guard let handle else { return call.reject("LanceDB not initialized — call open() first") }
        guard let key = call.getString("key") else { return call.reject("key is required") }

        Task {
            do {
                try await handle.delete(key: key)
                call.resolve()
            } catch {
                call.reject("memoryDelete failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func memoryList(_ call: CAPPluginCall) {
        guard let handle else { return call.reject("LanceDB not initialized — call open() first") }

        let prefix = call.getString("prefix")
        let limit = call.getInt("limit").map { UInt32($0) }

        Task {
            do {
                let keys = try await handle.list(prefix: prefix, limit: limit)
                call.resolve(["keys": keys])
            } catch {
                call.reject("memoryList failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func memoryClear(_ call: CAPPluginCall) {
        guard let handle else { return call.reject("LanceDB not initialized — call open() first") }

        let collection = call.getString("collection")

        Task {
            do {
                try await handle.clear(collection: collection)
                call.resolve()
            } catch {
                call.reject("memoryClear failed: \(error.localizedDescription)")
            }
        }
    }
}
